fdb-extensions/src/main/java/com/apple/foundationdb/async/hnsw/HNSW.java

```diff
@@ -0,0 +1,2022 @@
+     * <p>
+     * This search is performed on layers that use {@code InliningNode}s, where neighbor vectors are stored directly
+     * within the node.
+     * The search starts from a given {@code entryNeighbor} and iteratively moves to the closest neighbor in the current
+     * node's
+     * neighbor list, until no closer neighbor can be found.
+     * <p>
+     * The entire process is asynchronous, returning a {@link CompletableFuture} that will complete with the best node
+     * found in this layer.
+     *
+     * @param storageAdapter the storage adapter to fetch nodes from the graph
+     * @param readTransaction the transaction context for database reads
+     * @param storageTransform an affine transformation operator that is used to transform the fetched vector into the
+     * storage space that is currently being used
+     * @param estimator a distance estimator
+     * @param entryNeighbor the entry point for the search in this layer, typically the result from a search in a higher
+     * layer
+     * @param layer the layer number to perform the search in. Must be greater than 0.
+     * @param queryVector the vector for which to find the nearest neighbor
+     *
+     * @return a {@link CompletableFuture} that, upon completion, will hold the {@link NodeReferenceWithDistance} of the
+     * nearest neighbor found in this layer's greedy search
+     *
+     * @throws IllegalStateException if a node that is expected to exist cannot be fetched from the
+     * {@code storageAdapter} during the search
+     */
+    @Nonnull
+    private CompletableFuture<NodeReferenceWithDistance>
+            greedySearchInliningLayer(@Nonnull final StorageAdapter<NodeReferenceWithVector> storageAdapter,
+                                      @Nonnull final ReadTransaction readTransaction,
+                                      @Nonnull final AffineOperator storageTransform,
+                                      @Nonnull final Estimator estimator,
+                                      @Nonnull final NodeReferenceWithDistance entryNeighbor,
+                                      final int layer,
+                                      @Nonnull final RealVector queryVector) {
+        Verify.verify(layer > 0);
+        final AtomicReference<NodeReferenceWithDistance> currentNodeReferenceAtomic =
+                new AtomicReference<>(entryNeighbor);
+
+        return AsyncUtil.whileTrue(() -> onReadListener.onAsyncRead(
+                        storageAdapter.fetchNode(readTransaction, storageTransform, layer,
+                                currentNodeReferenceAtomic.get().getPrimaryKey()))
+                .thenApply(node -> {
+                    if (node == null) {
+                        throw new IllegalStateException("unable to fetch node");
+                    }
+                    final InliningNode inliningNode = node.asInliningNode();
+                    final List<NodeReferenceWithVector> neighbors = inliningNode.getNeighbors();
```

Comment:
It took me a bit, but if I understand correctly, the reason you need to have separate functions for `greedySearchInliningLayer` versus `searchLayer` is that with in-lining nodes, you can know the distances of all of the neighbors without needing to first do individual lookups. So, we don't need to keep a separate set of "candidate nodes with unknown distance" that `searchLayer` does.

I wonder if there's a way of collapsing those two cases a bit more. It's probably fine to leave it as separate methods, though